File: src/lib/catalog/wiki.ts
/**
 * Wiki data layer.
 *
 * Wiki pages are unified across three corpora (tables / docs / code).
 * Every operation is content-hash-gated so the agent's tick is a no-op when
 * nothing changed. Every create/update/skip is mirrored to `audit_log` so the
 * Activity feed shows a real trail of what the wiki agents did.
 */
import { createHash } from "node:crypto";
import { catalogPool } from "./db";
import { audit } from "./queries";

export type WikiKind = "tables" | "docs" | "code";

export type WikiPageInput = {
  kind: WikiKind;
  slug: string;
  title: string;
  summary?: string | null;
  body_md: string;
  source_ref?: unknown;
};

export type WikiPageRow = {
  id: number;
  kind: WikiKind;
  slug: string;
  title: string;
  summary: string | null;
  body_md: string;
  source_ref: unknown;
  content_hash: string;
  status: string;
  generated_at: string | null;
  last_seen_at: string;
  created_at: string;
  updated_at: string;
};

export function pageHash(body_md: string, source_ref: unknown): string {
  return createHash("md5")
    .update(body_md)
    .update("\n")
    .update(JSON.stringify(source_ref ?? null))
    .digest("hex");
}

/**
 * Upsert a wiki page with hash-gated update. Returns:
 *   { action: 'created' | 'updated' | 'skipped' }
 *
 * `skipped` means the content_hash matched; we touched last_seen_at so the
 * janitor doesn't think the page was abandoned.
 */
export async function upsertWikiPage(
  actor: string,
  input: WikiPageInput,
): Promise<{ action: "created" | "updated" | "skipped"; id: number }> {
  const hash = pageHash(input.body_md, input.source_ref);
  // Existing?
  const existing = await catalogPool.query<{ id: number; content_hash: string }>(
    `SELECT id, content_hash FROM wiki_pages WHERE kind = $1 AND slug = $2`,
    [input.kind, input.slug],
  );
  if (existing.rows[0]) {
    const row = existing.rows[0];
    if (row.content_hash === hash) {
      await catalogPool.query(`UPDATE wiki_pages SET last_seen_at = NOW() WHERE id = $1`, [row.id]);
      await audit(actor, "wiki:page_skip", `${input.kind}/${input.slug}`, { id: row.id, reason: "hash_match" });
      return { action: "skipped", id: row.id };
    }
    await catalogPool.query(
      `UPDATE wiki_pages
          SET title = $2,
              summary = $3,
              body_md = $4,
              source_ref = $5::jsonb,
              content_hash = $6,
              status = 'ready',
              generated_at = NOW(),
              last_seen_at = NOW(),
              updated_at = NOW()
        WHERE id = $1`,
      [row.id, input.title, input.summary ?? null, input.body_md, JSON.stringify(input.source_ref ?? null), hash],
    );
    await replaceLinks(row.id, input.body_md);
    await audit(actor, "wiki:page_update", `${input.kind}/${input.slug}`, {
      id: row.id,
      bytes: input.body_md.length,
    });
    return { action: "updated", id: row.id };
  }
  const r = await catalogPool.query<{ id: number }>(
    `INSERT INTO wiki_pages (kind, slug, title, summary, body_md, source_ref, content_hash, status, generated_at)
       VALUES ($1, $2, $3, $4, $5, $6::jsonb, $7, 'ready', NOW())
     RETURNING id`,
    [input.kind, input.slug, input.title, input.summary ?? null, input.body_md, JSON.stringify(input.source_ref ?? null), hash],
  );
  const id = r.rows[0]!.id;
  await replaceLinks(id, input.body_md);
  await audit(actor, "wiki:page_create", `${input.kind}/${input.slug}`, {
    id,
    bytes: input.body_md.length,
  });
  return { action: "created", id };
}

/**
 * Mark pages that haven't been seen for N hours as stale. The runner
 * touches last_seen_at on every tick, so a stale page is one whose
 * source disappeared.
 */
export async function markStaleWikiPages(actor: string, kind: WikiKind, olderThanHours: number = 24) {
  const r = await catalogPool.query<{ id: number; slug: string }>(
    `UPDATE wiki_pages
        SET status = 'stale'
      WHERE kind = $1 AND status = 'ready'
        AND last_seen_at < NOW() - ($2 || ' hours')::interval
     RETURNING id, slug`,
    [kind, String(olderThanHours)],
  );
  for (const row of r.rows) {
    await audit(actor, "wiki:page_stale", `${kind}/${row.slug}`, { id: row.id, hours: olderThanHours });
  }
  return r.rowCount ?? 0;
}

/**
 * Reset the agent's "in-progress" lock + write a status row. Called
 * around every wiki agent tick.
 */
export async function beginAgentTick(kind: WikiKind): Promise<boolean> {
  // Try to claim the soft lock. Returns true if claimed.
  const r = await catalogPool.query<{ kind: WikiKind }>(
    `UPDATE wiki_agent_state
        SET is_running = TRUE
      WHERE kind = $1 AND is_running = FALSE
     RETURNING kind`,
    [kind],
  );
  return (r.rowCount ?? 0) > 0;
}

export async function endAgentTick(
  kind: WikiKind,
  status: "ok" | "failed",
  pagesGenerated: number,
  error?: string,
): Promise<void> {
  await catalogPool.query(
    `UPDATE wiki_agent_state
        SET is_running = FALSE,
            last_run_at = NOW(),
            last_status = $2,
            last_error = $3,
            pages_generated = pages_generated + $4
      WHERE kind = $1`,
    [kind, status, error ?? null, pagesGenerated],
  );
}

async function replaceLinks(fromPageId: number, body_md: string): Promise<void> {
  await catalogPool.query(`DELETE FROM wiki_links WHERE from_page_id = $1`, [fromPageId]);
  const refs = parseWikiLinks(body_md);
  if (refs.length === 0) return;
  for (const r of refs) {
    await catalogPool.query(
      `INSERT INTO wiki_links (from_page_id, to_kind, to_slug)
         VALUES ($1, $2, $3)
       ON CONFLICT DO NOTHING`,
      [fromPageId, r.kind, r.slug],
    );
  }
}

/**
 * Cross-link syntax: [[tables/deviations]], [[docs/qa-runbook]], or just
 * [[deviations]] (defaults to current page's kind — but at parse time we
 * don't know it; require the explicit `kind/slug` form).
 */
export function parseWikiLinks(md: string): { kind: WikiKind; slug: string }[] {
  const out: { kind: WikiKind; slug: string }[] = [];
  const re = /\[\[(tables|docs|code)\/([a-z0-9][a-z0-9-_]{1,80})\]\]/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(md)) !== null) {
    out.push({ kind: m[1] as WikiKind, slug: m[2]! });
  }
  return out;
}

export async function listWikiPages(kind?: WikiKind): Promise<WikiPageRow[]> {
  const r = await catalogPool.query<WikiPageRow>(
    kind
      ? `SELECT id, kind, slug, title, summary, body_md, source_ref, content_hash, status, generated_at::text, last_seen_at::text, created_at::text, updated_at::text
           FROM wiki_pages WHERE kind = $1 ORDER BY title ASC`
      : `SELECT id, kind, slug, title, summary, body_md, source_ref, content_hash, status, generated_at::text, last_seen_at::text, created_at::text, updated_at::text
           FROM wiki_pages ORDER BY kind, title ASC`,
    kind ? [kind] : [],
  );
  return r.rows;
}

export async function getWikiPage(kind: WikiKind, slug: string): Promise<WikiPageRow | null> {
  const r = await catalogPool.query<WikiPageRow>(
    `SELECT id, kind, slug, title, summary, body_md, source_ref, content_hash, status, generated_at::text, last_seen_at::text, created_at::text, updated_at::text
       FROM wiki_pages WHERE kind = $1 AND slug = $2`,
    [kind, slug],
  );
  return r.rows[0] ?? null;
}

export async function getBacklinks(toKind: WikiKind, toSlug: string): Promise<WikiPageRow[]> {
  const r = await catalogPool.query<WikiPageRow>(
    `SELECT p.id, p.kind, p.slug, p.title, p.summary, p.body_md, p.source_ref, p.content_hash, p.status,
            p.generated_at::text, p.last_seen_at::text, p.created_at::text, p.updated_at::text
       FROM wiki_links l
       JOIN wiki_pages p ON p.id = l.from_page_id
      WHERE l.to_kind = $1 AND l.to_slug = $2
      ORDER BY p.title`,
    [toKind, toSlug],
  );
  return r.rows;
}

File: backend/app/catalog/wiki.py
"""Wiki data layer — hash-gated upserts + backlinks + agent-tick locks.

Port of src/lib/catalog/wiki.ts. Pages span three corpora (tables / docs / code).
Every write is mirrored to audit_log so the Activity feed shows agent actions.
"""
from __future__ import annotations

import hashlib
import json
import re
from typing import Any, Literal

from psycopg.rows import dict_row
from psycopg.types.json import Jsonb

from ..audit import audit
from ..db import get_conn

WikiKind = Literal["tables", "docs", "code"]

_LINK_RE = re.compile(r"\[\[(tables|docs|code)/([a-z0-9][a-z0-9\-_]{1,80})\]\]")
_COLS = (
    "id, kind, slug, title, summary, body_md, source_ref, content_hash, status, "
    "generated_at::text AS generated_at, last_seen_at::text AS last_seen_at, "
    "created_at::text AS created_at, updated_at::text AS updated_at"
)


def page_hash(body_md: str, source_ref: Any) -> str:
    h = hashlib.md5()
    h.update(body_md.encode("utf-8"))
    h.update(b"\n")
    h.update(json.dumps(source_ref if source_ref is not None else None, sort_keys=True).encode("utf-8"))
    return h.hexdigest()


def parse_wiki_links(md: str) -> list[dict[str, str]]:
    return [{"kind": m.group(1), "slug": m.group(2)} for m in _LINK_RE.finditer(md)]


async def upsert_wiki_page(
    actor: str,
    *,
    kind: WikiKind,
    slug: str,
    title: str,
    body_md: str,
    summary: str | None = None,
    source_ref: Any = None,
) -> dict[str, Any]:
    """Returns dict with keys: action ('created'|'updated'|'skipped'), id."""
    h = page_hash(body_md, source_ref)
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                "SELECT id, content_hash FROM wiki_pages WHERE kind = %s AND slug = %s",
                (kind, slug),
            )
            existing = await cur.fetchone()

    if existing:
        if existing["content_hash"] == h:
            async with get_conn() as conn:
                async with conn.cursor() as cur:
                    await cur.execute(
                        "UPDATE wiki_pages SET last_seen_at = NOW() WHERE id = %s",
                        (existing["id"],),
                    )
            await audit(actor, "wiki:page_skip", f"{kind}/{slug}",
                        {"id": existing["id"], "reason": "hash_match"})
            return {"action": "skipped", "id": int(existing["id"])}

        async with get_conn() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    """
                    UPDATE wiki_pages
                       SET title = %s, summary = %s, body_md = %s,
                           source_ref = %s, content_hash = %s,
                           status = 'ready', generated_at = NOW(),
                           last_seen_at = NOW(), updated_at = NOW()
                     WHERE id = %s
                    """,
                    (title, summary, body_md, Jsonb(source_ref), h, existing["id"]),
                )
        await _replace_links(int(existing["id"]), body_md)
        await audit(actor, "wiki:page_update", f"{kind}/{slug}",
                    {"id": existing["id"], "bytes": len(body_md)})
        return {"action": "updated", "id": int(existing["id"])}

    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                INSERT INTO wiki_pages
                    (kind, slug, title, summary, body_md, source_ref,
                     content_hash, status, generated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, 'ready', NOW())
                RETURNING id
                """,
                (kind, slug, title, summary, body_md, Jsonb(source_ref), h),
            )
            row = await cur.fetchone()
    assert row is not None
    new_id = int(row[0])
    await _replace_links(new_id, body_md)
    await audit(actor, "wiki:page_create", f"{kind}/{slug}",
                {"id": new_id, "bytes": len(body_md)})
    return {"action": "created", "id": new_id}


async def mark_stale_wiki_pages(actor: str, kind: WikiKind, older_than_hours: int = 24) -> int:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                UPDATE wiki_pages
                   SET status = 'stale'
                 WHERE kind = %s AND status = 'ready'
                   AND last_seen_at < NOW() - (%s || ' hours')::interval
                RETURNING id, slug
                """,
                (kind, str(older_than_hours)),
            )
            rows = await cur.fetchall()
    for row in rows:
        await audit(actor, "wiki:page_stale", f"{kind}/{row['slug']}",
                    {"id": row["id"], "hours": older_than_hours})
    return len(rows)


async def begin_agent_tick(kind: WikiKind) -> bool:
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                UPDATE wiki_agent_state
                   SET is_running = TRUE
                 WHERE kind = %s AND is_running = FALSE
                RETURNING kind
                """,
                (kind,),
            )
            row = await cur.fetchone()
    return row is not None


async def end_agent_tick(
    kind: WikiKind, status: str, pages_generated: int, error: str | None = None
) -> None:
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                """
                UPDATE wiki_agent_state
                   SET is_running = FALSE,
                       last_run_at = NOW(),
                       last_status = %s,
                       last_error = %s,
                       pages_generated = pages_generated + %s
                 WHERE kind = %s
                """,
                (status, error, pages_generated, kind),
            )


async def list_wiki_pages(kind: WikiKind | None = None) -> list[dict[str, Any]]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            if kind:
                await cur.execute(
                    f"SELECT {_COLS} FROM wiki_pages WHERE kind = %s ORDER BY title ASC",
                    (kind,),
                )
            else:
                await cur.execute(
                    f"SELECT {_COLS} FROM wiki_pages ORDER BY kind, title ASC"
                )
            return await cur.fetchall()


async def get_wiki_page(kind: WikiKind, slug: str) -> dict[str, Any] | None:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                f"SELECT {_COLS} FROM wiki_pages WHERE kind = %s AND slug = %s",
                (kind, slug),
            )
            return await cur.fetchone()


async def get_backlinks(to_kind: WikiKind, to_slug: str) -> list[dict[str, Any]]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                f"""
                SELECT {_COLS.replace('id, ', 'p.id AS id, ')
                              .replace('kind, ', 'p.kind AS kind, ')
                              .replace('slug, ', 'p.slug AS slug, ')
                              .replace('title, ', 'p.title AS title, ')
                              .replace('summary, ', 'p.summary AS summary, ')
                              .replace('body_md, ', 'p.body_md AS body_md, ')
                              .replace('source_ref, ', 'p.source_ref AS source_ref, ')
                              .replace('content_hash, ', 'p.content_hash AS content_hash, ')
                              .replace('status, ', 'p.status AS status, ')
                              .replace('generated_at::text AS generated_at', 'p.generated_at::text AS generated_at')
                              .replace('last_seen_at::text AS last_seen_at', 'p.last_seen_at::text AS last_seen_at')
                              .replace('created_at::text AS created_at', 'p.created_at::text AS created_at')
                              .replace('updated_at::text AS updated_at', 'p.updated_at::text AS updated_at')}
                  FROM wiki_links l
                  JOIN wiki_pages p ON p.id = l.from_page_id
                 WHERE l.to_kind = %s AND l.to_slug = %s
                 ORDER BY p.title
                """,
                (to_kind, to_slug),
            )
            return await cur.fetchall()


async def _replace_links(from_page_id: int, body_md: str) -> None:
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute(
                "DELETE FROM wiki_links WHERE from_page_id = %s", (from_page_id,)
            )
            for ref in parse_wiki_links(body_md):
                await cur.execute(
                    """
                    INSERT INTO wiki_links (from_page_id, to_kind, to_slug)
                         VALUES (%s, %s, %s)
                    ON CONFLICT DO NOTHING
                    """,
                    (from_page_id, ref["kind"], ref["slug"]),
                )
